concurrency/PostgresMultiWarehouseDataSource.java

```diff
@@ -0,0 +1,221 @@
+                log.error(errorMessage);
+                throw new IllegalStateException(errorMessage);
+            }
+            log.info("Creating empty postgres datasource for application initialization.");
+            return new DriverManagerDataSource();
+        }
+        return getPostgresDataSourceForWarehouseId(warehouseId);
+    }
+
+    private DataSource getPostgresDataSourceForWarehouseId(String warehouseId) throws CryptoException, PropertyVetoException
+    {
+        synchronized (this)
+        {
+            if (dataSourceConnectionCache.containsKey(warehouseId))
+            {
+                log.debug("Found a cached DataSource for warehouseId: {}", warehouseId);
+                return dataSourceConnectionCache.get(warehouseId);
+            }
+            else
+            {
+                log.info("Creating new DataSource for warehouseId: {}", warehouseId);
+
+                OdinAWSCredentialsProvider access = new OdinAWSCredentialsProvider(buildOdinMaterialSetName(warehouseId));
+
+                Properties props = new Properties();
+                props.setProperty(PROPERTY_NAME_USERNAME, access.getCredentials().getAWSAccessKeyId());
+                props.setProperty(PROPERTY_NAME_PASSWORD, decryptOdinPassword(access.getCredentials().getAWSSecretKey()));
+                props.setProperty(PROPERTY_NAME_SSL_ENABLED, PROPERTY_VALUE_SSL_ENABLED);
+                props.setProperty(PROPERTY_NAME_SSL_MODE, PROPERTY_VALUE_SSL_MODE);
+                props.setProperty(PROPERTY_NAME_SSL_CERT_PATH, PROPERTY_VALUE_SSL_CERT_PATH);
+                props.setProperty(PROPERTY_NAME_APPLICATION_NAME, PROPERTY_VALUE_APPLICATION_NAME);
+
+                String connectionUrl = buildJdbcUrl(warehouseId);
+                ComboPooledDataSource dataSource = new ComboPooledDataSource();
+                dataSource.setDriverClass(POSTGRES_DRIVER_CLASS);
+                dataSource.setProperties(props);
+                dataSource.setJdbcUrl(connectionUrl);
+
+                dataSourceConnectionCache.put(warehouseId, dataSource);
+                log.info("Successfully created new Postgres DataSource for warehouseId: {} and placed into local cache", warehouseId);
+                return dataSource;
+            }
+        }
+    }
+
+    private String decryptOdinPassword(String encryptedPassword) throws CryptoException
+    {
+        OdinPasswordStore odinPasswordStore = new OdinPasswordStore();
+        String decryptedPassword = odinPasswordStore.decryptPassword(encryptedPassword);
+        log.debug("Successfully decrypted password");
+        return decryptedPassword;
+    }
+
+    @Override
+    public Connection getConnection() throws SQLException
+    {
+        DataSource dataSource = null;
+        try
+        {
+            dataSource = resolveTargetDataSource();
+        }
+        catch (CryptoException | PropertyVetoException e)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)